fc/parsing/actions.py

```diff
@@ -780,6 +864,99 @@ def sympy_equations(self):
                 eqs.append(eq.to_sympy(self._symbol_generator, self._number_generator))
         return self._sympy_equations
 
+    #######################################
+    # Helper methods for model manipulation
+
+    def _convert_time_if_needed(self):
+        """Units-convert the time variable if not in the protocol's units."""
+        time_units = self.units.get_unit(self.time_units)
+        time_var = self.model.get_free_variable_symbol()
+        time_var = self.model.convert_variable(time_var, time_units, DataDirectionFlow.INPUT)
+
+    def _add_input_variables(self):
+        """Ensure input variables exist in the desired units.
+
+        TODO: At present this just checks they exist; no units conversion or initial_value handling.
+        TODO: If an input doesn't exist but has units & initial_value here, or is defined via
+        an equation, then add a new variable to the model.
+        """
+        for var in self.inputs:
+            assert var.units is None
+            assert var.initial_value is None
+            self.model.get_symbol_by_ontology_term(var.rdf_term)
+
+    def _add_or_replace_equations(self):
+        """Process define statements and modify the model's equations accordingly."""
+        for eq in self.sympy_equations:
+            lhs = eq.lhs
+            if lhs.is_Derivative:
+                var = lhs.args[0]
+                assert var.initial_value is not None  # TODO: Maybe throw ProtocolError instead?
+            else:
+                var = lhs
+                var.initial_value = None  # In case it was a state variable previously
+            # Figure out if this is a replace or add
+            defn = self.model.get_definition(var)
+            if defn is not None:
+                self.model.remove_equation(defn)
+            self.model.add_equation(eq)
+        # TODO: Check units of newly added equations; apply conversions where needed?
```

Comment:
Maybe already add the check, but raise some sort of NotImplementedError? That way we'll know where to look when we start implementing :D

Reply:
I've stuck TODO comments in for everything not handled yet, so we can just search for those. Not sure yet exactly what the check should look like!

Reply:
Fair enough!